File: src/main/java/com/github/sarhatabaot/messages/WriteYamlClass.java
package com.github.sarhatabaot.messages;

import org.jetbrains.annotations.NotNull;
import org.yaml.snakeyaml.Yaml;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.util.Map;
import java.util.Set;

/**
 * @author sarhatabaot
 */
public class WriteYamlClass extends WriteClass<Object>{
    public WriteYamlClass(final String targetPackage, final String basePath, final String privateConstructor, final boolean overwriteClasses) {
        super(targetPackage, basePath, privateConstructor, overwriteClasses);
    }

    @Override
    public FileType getFileType() {
        return FileType.YAML;
    }

    @Override
    public boolean isPrimitive(final @NotNull Object element) {
        return element.getClass().isPrimitive() ||
                element instanceof Integer ||
                element instanceof Double ||
                element instanceof Float ||
                element instanceof Boolean ||
                element instanceof String;
    }

    @Override
    public Set<Map.Entry<String, Object>> getRootEntrySet(final File file) throws IOException {
        Yaml yaml = new Yaml();
        try (InputStream inputStream = Files.newInputStream(file.toPath())) {
            Map<String, Object> yamlFile = yaml.load(inputStream);
            return yamlFile.entrySet();
        }
    }

    @SuppressWarnings("unchecked")
    @Override
    public Set<Map.Entry<String, Object>> getEntrySetFromValue(final Object value) {
        return ((Map<String,Object>) value).entrySet();
    }

    @Override
    public TypeKeyValue getEntryValue(final Object value) {
        if(value instanceof Integer)
            return new TypeKeyValue(Integer.class, String.valueOf(value));
        if(value instanceof Double)
            return new TypeKeyValue(Double.class, String.valueOf(value));
        if(value instanceof Boolean)
            return new TypeKeyValue(Boolean.class,String.valueOf(value));

        return new TypeKeyValue(String.class, String.valueOf(value));
    }

    @Override
    public boolean isArray(final Object element) {
        return element instanceof String[];
    }

    @Override
    public String[] getAsStringArray(final Object value) {
        return (String[]) value;
    }
}

File: src/main/kotlin/com/github/sarhatabaot/messages/generate/WriteYamlClass.kt
package com.github.sarhatabaot.messages.generate

import com.github.sarhatabaot.messages.model.FileType
import com.github.sarhatabaot.messages.model.TypeKeyValue
import org.yaml.snakeyaml.Yaml
import java.io.File
import java.nio.file.Files


/**
 *
 * @author sarhatabaot
 */
class WriteYamlClass(
    targetPackage: String,
    basePath: String,
    privateConstructor: String?,
    overwriteClasses: Boolean
): WriteClass<Any>(
    targetPackage,
    basePath,
    privateConstructor,
    overwriteClasses
) {
    override fun getFileType() = FileType.YAML

    override fun isPrimitive(element: Any): Boolean {
        return element.javaClass.isPrimitive ||
                element is Int ||
                element is Double ||
                element is Float ||
                element is Boolean ||
                element is String
    }

    override fun isArray(element: Any): Boolean {
        return element is Array<*> && element.isArrayOf<String>()
    }

    override fun getRootEntrySet(file: File?): Set<Map.Entry<String, Any>> {
        val yaml = Yaml()
        Files.newInputStream(file!!.toPath()).use { inputStream ->
            val yamlFile =
                yaml.load<Map<String, Any>>(inputStream)
            return yamlFile.entries
        }
    }

    override fun getEntrySetFromValue(value: Any) = (value as Map<String, Any>).entries

    override fun getEntryValue(value: Any): TypeKeyValue {
        if (value is Int) return TypeKeyValue(Int::class.java, value.toString())
        if (value is Double) return TypeKeyValue(Double::class.java, value.toString())
        return if (value is Boolean) TypeKeyValue(
            Boolean::class.java,
            value.toString()
        ) else TypeKeyValue(String::class.java, value.toString())

    }

    override fun getAsStringArray(value: Any): Array<String> {
        return value as Array<String>
    }
}